x-pack/plugins/security_solution/server/lib/detection_engine/routes/fleet/get_installed_integrations/get_installed_integrations_route.ts

```diff
@@ -39,17 +46,26 @@ export const getInstalledIntegrationsRoute = (router: SecuritySolutionPluginRout
           set.addPackagePolicy(policy);
         });
 
-        const registryPackages = await Promise.all(
-          set.getPackages().map((packageInfo) => {
-            return fleet.packages.getRegistryPackage(
+        const registryPackages = await initPromisePool({
+          concurrency: MAX_CONCURRENT_REQUESTS_TO_PACKAGE_REGISTRY,
```

Comment:
> Would be good to test with a large number of integration policies, and then also with a large number of installed integrations.

Agree, will do it 👍 

> One issue we have with fetching package policies is we're not [handling](https://github.com/elastic/kibana/blob/58bb62cd4d8f2beea7bdad1500a06f123122b6d3/x-pack/plugins/security_solution/server/lib/detection_engine/routes/fleet/get_installed_integrations/get_installed_integrations_route.ts#L43) [pagination](https://github.com/elastic/kibana/blob/3730dd0779ed8efd74aee88f57422781ec1ac122/x-pack/plugins/fleet/common/types/rest_spec/common.ts#L21-L22), so we'll cap out at the max for that initial page.

I'll check how this method works when the `perPage` is not specified. I thought it just returns all the policies, but it might also set a default page size instead. Anyway, definitely need to test this 👍 

> I'm a little worried if there's say 100+ installed packages and we end up making 100+ requests to fleet each time an individual user hits the Rules Mgmt or Details pages. We do have caching client-side through the react-query wrapper, but nothing here server side

FWIW Fleet service caches fetched packages on the server-side. So only the first request may be slow (it depends linearly on the number of installed packages). The subsequent requests should be fast and not generate outgoing HTTP requests to EPR. I'm not sure 100% about this though -- so will test it as well. 👍 

I will also test it with a large number of installed packages and a large number of integration policies.

> Will be good to get feedback from the fleet folks about any other API's we might be able to use here to ease the pressure on the fleet side -- would be great if we could just get all installed packages in one request instead of going the integration policy route.

I'll open a discussion issue and share it with Fleet folks. 👍 

Thank you @spong, this is a lot of great points!

Reply:
> FWIW Fleet service caches fetched packages on the server-side. So only the first request may be slow (it depends linearly on the number of installed packages). The subsequent requests should be fast and not generate outgoing HTTP requests to EPR. I'm not sure 100% about this though -- so will test it as well. 👍

This makes me feel a lot better about the upper bound cases. I didn't realize the fleet service cached as well, thanks @banderror! 

And sounds good with the other points as well, thank you for verifying here 👍